batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+	// Send the image to the result channel
+	defer wg.Done()
+	for {
+		row, ok := <-p.channels[CONVERT]
+		if !ok {
+			break
+		}
+
+		if row.Err != nil {
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		file, err := os.Open(row.Output)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		// Upload to S3
+		path := strings.Split(row.Output, "/")
+		key := path[len(path)-1]
+
+		err = UploadToS3WithBackoff(file, key, p.config.Aws, p.maxRetries)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		row.S3url = fmt.Sprintf("https://%s.s3.amazonaws.com/%s", p.config.Aws.s3bucket, key)
+
+		p.channels[UPLOAD] <- row
+	}
+}
+
+func (p *Pipeline) Write(done chan bool) {
+
+	f, err := os.Create(p.config.OutputFilepath)
+	if err != nil {
+		log.Fatalf("error creating output file: %v", err)
+	}
+	defer f.Close()
+
+	w := csv.NewWriter(f)
+	defer w.Flush()
+	w.Write(OutputHeader)
+	var failedW *csv.Writer
+
+	// Create csv for failed output if param is set
+	if p.config.FailedOutputFilepath != "" {
+		failedF, err := os.Create(p.config.FailedOutputFilepath)
+		if err != nil {
+			log.Fatalf("error creating failed output file: %v", err)
+		}
+		defer failedF.Close()
+
+		failedW = csv.NewWriter(failedF)
+		defer failedW.Flush()
+		failedW.Write(FailedOutputHeader)
+	}
+
+	for {
+		row, ok := <-p.channels[UPLOAD]
+
+		if !ok {
+			break
+		}
+		rowErr := ""
+		// only add to failed output if row has an error, but not a duplicate error or image exists error
+		if row.Err != nil {
+			if p.config.FailedOutputFilepath != "" && row.Err != ErrDuplicateURL && row.Err != ErrImageExistsInS3 {
+				failedW.Write([]string{row.Url})
+			}
+			rowErr = row.Err.Error()
+		}
+
+		w.Write([]string{row.Url, row.Input, row.Output, row.S3url, rowErr})
+	}
+	done <- true
+}
+
+func (p *Pipeline) taskPicker(task string) Task {
+	switch task {
+	case DOWNLOAD:
+		return p.Download
+	case CONVERT:
+		return p.Convert
+	case UPLOAD:
+		return p.Upload
+	}
+
+	return nil
+}
+
+func (p *Pipeline) Execute() error {
+
+	start := time.Now()
+
+	done := make(chan bool)
+	go p.Write(done)
+
+	// Start the workers
+	wgTasks := &sync.WaitGroup{}
+	for _, task := range TASKS {
+		wgTasks.Add(1)
+		go func(task string) {
+			var wg sync.WaitGroup
+			for i := 0; i < p.workers[task]; i++ {
+				wg.Add(1)
+				go p.taskPicker(task)(&wg)
+			}
+			wg.Wait()
+			close(p.channels[task])
+			wgTasks.Done()
+		}(task)
+	}
+
+	// Start the reader
+	csvReader, err := OpenCSVFile(p.config.InputFilepath)
+
+	if err != nil {
+		return err
+	}
+	go p.Read(csvReader)
+
+	// Wait for all tasks to finish
+	wgTasks.Wait()
+	elapsed := time.Since(start)
+
+	log.Printf("Finished in %s", elapsed)
+
+	<-done
+
+	return nil
+}
+
+func NewPipeline(config *Config) *Pipeline {
+	return &Pipeline{
+		config:     config,
+		maxRetries: 3,
+		urlMap:     &sync.Map{},
+		channels: map[string]chan *Out{
+			READ:     make(chan *Out),
```

Comment:
What trade-offs are you making by having these channels be un-buffered? What would the pros and cons be to making them buffered? If you made them buffered, how big a buffer do you think would make sense?

Reply:
When making channels unbuffered there's a big chance that some jobs will have to wait for resource intensive jobs to unblock the channel. If there are multiple intensive stages, it makes sense to have unbuffered channel as it would act as kind of a natural throttle.

If implementing buffered channels I'd think about bottlenecks, If my reads are being blocked because the channel is blocked - I'd consider setting a larger buffer since it's probably better to keep it in memory(if it's reasonable amount) than have a goroutine that's blocked.

Reply:
At the moment you have 10 download channels and 3 convert channels - what's the biggest number of parallel downloads you expect to happen?

After you've made that prediction, can you add some instrumentation to check the maximum number of downloads you _actually_ do in parallel?

(I'd recommend using the [`sync/atomic`](https://pkg.go.dev/sync/atomic#AddInt32) package to increment a counter at the start of each download, and decrement the counter at the end of each download, and somehow track / print the highest that number ever hits).